src/adapter/src/optimize/dataflow_expiration.rs

```diff
@@ -0,0 +1,212 @@
+            time_dependencies.retain(|dep| matches!(dep, RefreshSchedule(_, _)));
+            RefreshSchedule(schedule, time_dependencies)
+        } else {
+            // No wall-clock dependence, no refresh schedule
+            Indeterminate
+        };
+        time_dependence.normalize();
+        time_dependence
+    }
+
+    /// Determine the time dependence for a [`DataSourceDesc`].
+    ///
+    /// Non-epoch timelines are indeterminate, and load generators.
+    fn for_data_source_desc(
+        &mut self,
+        desc: &DataSourceDesc,
+        timeline: &Timeline,
+    ) -> TimeDependence {
+        use TimeDependence::*;
+        // We only know how to handle the epoch timeline.
+        if !matches!(timeline, Timeline::EpochMilliseconds) {
+            return Indeterminate;
+        }
+        match desc {
+            DataSourceDesc::Ingestion { ingestion_desc, .. } => {
+                match ingestion_desc.desc.connection {
+                    // Kafka, Postgres, MySql sources follow wall clock.
+                    GenericSourceConnection::Kafka(_)
+                    | GenericSourceConnection::Postgres(_)
+                    | GenericSourceConnection::MySql(_) => Wallclock,
+                    // Load generators not further specified.
+                    GenericSourceConnection::LoadGenerator(_) => Indeterminate,
+                }
+            }
+            DataSourceDesc::IngestionExport { ingestion_id, .. } => {
+                self.determine_dependence_inner(*ingestion_id)
+            }
+            // Introspection, progress and webhook sources follow wall clock.
+            DataSourceDesc::Introspection(_)
+            | DataSourceDesc::Progress
+            | DataSourceDesc::Webhook { .. } => Wallclock,
+        }
+    }
+
+    /// Determine the time dependence for a single global ID.
+    fn determine_dependence_inner(&mut self, id: GlobalId) -> TimeDependence {
+        use TimeDependence::*;
+
+        if let Some(dependence) = self.seen.get(&id).cloned() {
+            return dependence;
+        }
+
+        let entry = self.catalog.get_entry(&id);
+        let mut time_dependence = match entry.item_type() {
+            CatalogItemType::Table => {
+                if let Some(data_source) = entry.table().map(|table| &table.data_source) {
+                    match data_source {
+                        // Tables follow wall clock.
+                        TableDataSource::TableWrites { .. } => Wallclock,
+                        TableDataSource::DataSource { desc, timeline } => {
+                            self.for_data_source_desc(desc, timeline)
+                        }
+                    }
+                } else {
+                    Indeterminate
+                }
+            }
+            CatalogItemType::Source => {
+                // Some sources don't have a `Source` entry, so we can't determine their time
+                // dependence.
```

Comment:
Ah, perhaps this is why you're using `item_type()`. I see that both `Source` and `Log` are categorized as `Source`. 

I think we should capture this explicitly by matching on `entry.item()`. Will simply the code as well.